modules/rpm-ostree/rpm-ostree.sh

```diff
@@ -43,33 +47,101 @@ fi
 get_yaml_array INSTALL '.install[]' "$1"
 get_yaml_array REMOVE '.remove[]' "$1"
 
+CLASSIC_INSTALL=false
+HTTPS_INSTALL=false
+LOCAL_INSTALL=false
+
+# Install and remove RPM packages
+# Sort classic, URL & local packages
 if [[ ${#INSTALL[@]} -gt 0 ]]; then
-    for PKG in "${INSTALL[@]}"; do
-        if [[ "$PKG" =~ ^https?:\/\/.* ]]; then
-            REPLACED_PKG="${PKG//%OS_VERSION%/${OS_VERSION}}"
-            echo "Installing directly from URL: ${REPLACED_PKG}"
-            rpm-ostree install "$REPLACED_PKG"
-            INSTALL=( "${INSTALL[@]/$PKG}" ) # delete URL from install array
-        fi
-    done
+  for PKG in "${INSTALL[@]}"; do
+      if [[ "$PKG" =~ ^https?:\/\/.* ]]; then
+        VERSION_SUBSTITUTED_PKG="${PKG//%OS_VERSION%/${OS_VERSION}}"    
+        HTTPS_INSTALL=true
+        HTTPS_PKG+=("${VERSION_SUBSTITUTED_PKG}")
+      elif [[ ! "$PKG" =~ ^https?:\/\/.* ]] && [[ -f "${CONFIG_DIRECTORY}/rpm-ostree/${PKG}" ]]; then
+        LOCAL_INSTALL=true
+        LOCAL_PKG+=("${CONFIG_DIRECTORY}/rpm-ostree/${PKG}")
+      else
+        CLASSIC_INSTALL=true
+        CLASSIC_PKG+=("${PKG}")
+      fi
+  done
 fi
 
 # The installation is done with some wordsplitting hacks
 # because of errors when doing array destructuring at the installation step.
 # This is different from other ublue projects and could be investigated further.
-INSTALL_STR=$(echo "${INSTALL[*]}" | tr -d '\n')
+if ${CLASSIC_INSTALL} && ! ${HTTPS_INSTALL} && ! ${LOCAL_INSTALL}; then
+  INSTALL_STR=$(echo "${CLASSIC_PKG[*]}" | tr -d '\n')
+elif ! ${CLASSIC_INSTALL} && ${HTTPS_INSTALL} && ! ${LOCAL_INSTALL}; then
+  INSTALL_STR=$(echo "${HTTPS_PKG[*]}" | tr -d '\n')
+elif ! ${CLASSIC_INSTALL} && ! ${HTTPS_INSTALL} && ${LOCAL_INSTALL}; then
+  INSTALL_STR=$(echo "${LOCAL_PKG[*]}" | tr -d '\n')
+elif ${CLASSIC_INSTALL} && ${HTTPS_INSTALL} && ! ${LOCAL_INSTALL}; then
+  INSTALL_STR=$(echo "${CLASSIC_PKG[*]}" "${HTTPS_PKG[*]}" | tr -d '\n')
+elif ${CLASSIC_INSTALL} && ! ${HTTPS_INSTALL} && ${LOCAL_INSTALL}; then
+  INSTALL_STR=$(echo "${CLASSIC_PKG[*]}" "${LOCAL_PKG[*]}" | tr -d '\n')
+elif ! ${CLASSIC_INSTALL} && ${HTTPS_INSTALL} && ${LOCAL_INSTALL}; then
+  INSTALL_STR=$(echo "${HTTPS_PKG[*]}" "${LOCAL_PKG[*]}" | tr -d '\n')
+elif ${CLASSIC_INSTALL} && ${HTTPS_INSTALL} && ${LOCAL_INSTALL}; then
+  INSTALL_STR=$(echo "${CLASSIC_PKG[*]}" "${HTTPS_PKG[*]}" "${LOCAL_PKG[*]}" | tr -d '\n')
+fi
+
```

Comment:
This repetition feels very hacky, is it not possible to combine the package lists without having to check each combination?

Reply:
I agree.

I tried with case statements, but it still looked hacky.

Reply:
I tried some tests to simplify this, but for some reason, `rpm-ostree` would start to not do anything to install/remove RPMs in my test image.

What do you think of giving this a pass, while in the future, this should be refactored to process things fully as an array from beginning to the end, as opposed to this mess?

Reply:
Can you not concatenate arrays in bash, and only turn into string afterwards?

A quick googled search gave this https://stackoverflow.com/questions/31143874/how-to-concatenate-arrays-in-bash

Reply:
I made this all into arrays now & made the code more explainable.

It's much clearer now.

I don't know what was the issue that required "wordsplitting hacks", but whatever, that's gone now.

Reply:
I also don't like switching from array to string & vice-versa, I prefer to stay in 1 variable type & that's it.
Install & remove list is an array, so it's better to just keep it to be an array.

It prevents wordsplitting issues like those before this PR.